File: app/src/main/java/com/bachors/iptv/adapter/ChannelsAdapter.java
package com.bachors.iptv.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bachors.iptv.R;
import com.bachors.iptv.models.ChannelsData;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class ChannelsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<ChannelsData> allData;
    private Context inContext;

    public ChannelsAdapter(Context context) {
        this.inContext = context;
        allData = new ArrayList<>();
    }

    public List<ChannelsData> getData() {
        return allData;
    }

    public void setData(List<ChannelsData> allData) {
        this.allData = allData;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder = null;
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        viewHolder = getViewHolder(parent, inflater);
        return viewHolder;
    }

    @NonNull
    private RecyclerView.ViewHolder getViewHolder(ViewGroup parent, LayoutInflater inflater) {
        RecyclerView.ViewHolder viewHolder;
        View v1 = inflater.inflate(R.layout.item_channels, parent, false);
        viewHolder = new NisnVH(v1);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, final int position) {

        ChannelsData data = allData.get(position);

        final NisnVH nisnVH = (NisnVH) holder;

        String logo = data.getLogo();
        if(logo.equals("")){
            nisnVH.tvLogo.setImageDrawable(inContext.getResources().getDrawable(R.drawable.load));
        }else {
            nisnVH.tvName.setText(data.getName());
            Picasso.with(inContext)
                    .load(logo)
                    .placeholder(inContext.getResources().getDrawable(R.drawable.load))
                    .into(nisnVH.tvLogo);
        }

    }

    @Override
    public int getItemCount() {
        return allData == null ? 0 : allData.size();
    }

    /*
   Helpers
   _________________________________________________________________________________________________
    */


    public void add(ChannelsData r) {
        allData.add(r);
        notifyItemInserted(allData.size() - 1);
    }

    public void addAll(List<ChannelsData> semuaData) {
        for (ChannelsData data : semuaData) {
            add(data);
        }
    }

    public void remove(ChannelsData r) {
        int position = allData.indexOf(r);
        if (position > -1) {
            allData.remove(position);
            notifyItemRemoved(position);
        }
    }

    public void clear() {
        while (getItemCount() > 0) {
            remove(getItem(0));
        }
    }

    public boolean isEmpty() {
        return getItemCount() == 0;
    }

    public ChannelsData getItem(int position) {
        return allData.get(position);
    }


   /*
   View Holders
   _________________________________________________________________________________________________
    */

    /**
     * Main list's content ViewHolder
     */
    protected class NisnVH extends RecyclerView.ViewHolder {
        private TextView tvName;
        private ImageView tvLogo;

        public NisnVH(View itemView) {
            super(itemView);

            tvName = (TextView) itemView.findViewById(R.id.name);
            tvLogo = (ImageView) itemView.findViewById(R.id.logo);

        }
    }

}

File: app/src/main/java/com/bachors/iptv/adapters/ChannelsAdapter.kt
package com.bachors.iptv.adapters

import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.bachors.iptv.PlayerActivity
import com.bachors.iptv.R
import com.bachors.iptv.models.ChannelsData
import com.bachors.iptv.utils.SharedPrefManager
import com.squareup.picasso.Picasso
import org.json.JSONArray
import org.json.JSONObject

class ChannelsAdapter(private val inContext: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val allData = mutableListOf<ChannelsData>()
    private lateinit var sharedPrefManager: SharedPrefManager

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val v1 = inflater.inflate(R.layout.item_channels, parent, false)
        return ViewHolder(v1)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val data = allData[position]
        val holder = holder as ViewHolder

        sharedPrefManager = SharedPrefManager(inContext)

        val logo = data.logo
        holder.tvName.text = data.name
        Picasso.get()
            .load(logo)
            .placeholder(ContextCompat.getDrawable(inContext, R.drawable.load)!!)
            .into(holder.tvLogo)

        holder.lnPlay.setOnClickListener {
            val intent = Intent(inContext, PlayerActivity::class.java)
            intent.putExtra("name", data.name)
            intent.putExtra("url", data.url)
            inContext.startActivity(intent)
        }

        holder.btFavorite.setOnClickListener {
            try {
                val ar = JSONArray(sharedPrefManager.getSpFavorites())
                val ob = JSONObject()
                ob.put("name", data.name)
                ob.put("logo", data.logo)
                ob.put("url", data.url)
                ar.put(ob)
                sharedPrefManager.saveSPString(SharedPrefManager.SP_FAVORITES, ar.toString())
                Toast.makeText(inContext, "Saved to Favorites...", Toast.LENGTH_SHORT).show()
            } catch (_: Exception) {
                // Handle exception
            }
        }
    }

    override fun getItemCount(): Int = allData.size

    fun add(r: ChannelsData) {
        allData.add(r)
        notifyItemInserted(allData.size - 1)
    }

    fun addAll(semuaData: List<ChannelsData>) {
        for (data in semuaData) {
            add(data)
        }
    }

    fun remove(r: ChannelsData) {
        val position = allData.indexOf(r)
        if (position > -1) {
            allData.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun clear() {
        while (itemCount > 0) {
            remove(getItem(0))
        }
    }

    fun isEmpty(): Boolean = itemCount == 0

    fun getItem(position: Int): ChannelsData = allData[position]

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val lnPlay: LinearLayout = itemView.findViewById(R.id.play)
        val tvName: TextView = itemView.findViewById(R.id.name)
        val tvLogo: ImageView = itemView.findViewById(R.id.logo)
        val btFavorite: ImageView = itemView.findViewById(R.id.btn_favorite)
    }
}